tests/unit/src/main/scala/tests/TestingServer.scala

```diff
@@ -799,6 +800,42 @@ final class TestingServer(
     }
   }
 
+  def testClasses(filename: String): Future[List[TestDiscovery]] = {
+    val path = toPath(filename)
+    var retries = 5
+    val testClasses = Promise[List[TestDiscovery]]()
+    val handler = { refreshCount: Int =>
+      if (refreshCount > 0)
+        Thread.sleep(300)
+      executeCommand(ServerCommands.DiscoverTests)
+        .asInstanceOf[Future[ju.List[TestDiscovery]]]
+        .map(_.asScala.toList)
+        .foreach { r =>
+          if (r.exists(_.discovered.asScala.exists(_.nonEmpty))) {
+            testClasses.trySuccess(r)
+          } else if (retries > 0) {
+            retries -= 1
+            server.compilations.compileFile(path)
+          } else {
+            val error =
+              s"Could not fetch any test classes in $refreshCount tries"
+            testClasses.tryFailure(new NoSuchElementException(error))
+          }
+        }
+    }
+
+    Thread.sleep(300)
```

Comment:
Without those `Thread.sleep`s sometimes even 10 attempts was not enough and result was empty :/

Reply:
I had similar issues in the past with the sbt server suite and endup adding the `waitFor` method for this exact purpose, but instead used it in the for comprehension of the actual spec.

Reply:
`server.waitFor(TimeUnit.SECONDS.toMillis(10))` seems working, we'll see if it'll work in the CI. Thanks!